### PR TITLE
fix: [climate] Allow substitutions in `visual.temperature_step.{target_temperature,current_temperature}`

### PR DESCRIPTION
**Pull request by hostcc**
# What does this implement/fix?

The change addresses the error being raised when using substitutions in `visual.temperature_step`, `target_temperature` and `current_temperature` properties (see below for details on the error).

Basically, `visual.temperature_step` is now attempted as schema of separate values first, and then falls back to single value for both. Previously, the single value case has been processed first for schema interpretation and skipped `visual_temperature()` call on those, hence substitutions remain strings not floats.

	src/main.cpp: In function 'void setup()':
	src/main.cpp:135:85: error: no matching function for call to 'esphome::thermostat::ThermostatClimate::set_visual_temperature_step_override(const char [4], const char [4])'
	   thermostat_thermostatclimate_id->set_visual_temperature_step_override("0.5", "0.1");
											     ^
	In file included from src/esphome/components/climate/automation.h:4,
			 from src/esphome.h:3,
			 from src/main.cpp:3:
	src/esphome/components/climate/climate.h:250:8: note: candidate: 'void esphome::climate::Climate::set_visual_temperature_step_override(float, float)'
	   void set_visual_temperature_step_override(float target, float current);
		^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
	src/esphome/components/climate/climate.h:250:8: note:   no known conversion for argument 1 from 'const char [4]' to 'float'
	*** [.pioenvs/componenttestesp32ard/src/main.cpp.o] Error 1

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
substitutions:
  visual_target_temperature_step: '0.5'
  visual_current_temperature_step: '0.1'

sensor:
  - platform: template
    id: template_sensor

climate:
  - platform: thermostat
    name: Thermostat
    sensor: template_sensor
    min_idle_time: 0s
    idle_action: {}
    heat_action: {}
    min_heating_off_time: 0s
    min_heating_run_time: 0s
    visual:
      temperature_step:
        target_temperature: ${visual_target_temperature_step}
        current_temperature: ${visual_current_temperature_step}
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by ssieb** [comment]
Please refactor it so it's not recursive.

**Review by jesserockz — DISMISSED** [review]

**Comment by ssieb** [comment]
This is what I mean:
```python
VISUAL_TEMPERATURE_STEP_SCHEMA = cv.Schema(
        {
            cv.Required(CONF_TARGET_TEMPERATURE): visual_temperature,
            cv.Required(CONF_CURRENT_TEMPERATURE): visual_temperature,
        }
    )

def visual_temperature_steps(value):
    if isinstance(value, dict):
        return VISUAL_TEMPERATURE_STEP_SCHEMA(value)

    return VISUAL_TEMPERATURE_STEP_SCHEMA(
        {
            CONF_TARGET_TEMPERATURE: value,
            CONF_CURRENT_TEMPERATURE: value,
        }
    )

                    cv.Optional(CONF_TEMPERATURE_STEP): visual_temperature_steps,
```

It's a quick copy and paste and will need indent fixing, but that's the idea.

File: esphome/components/climate/__init__.py [hunk 1]
```diff
@@ -119,10 +119,19 @@
...
 
 
-def single_visual_temperature(value):
+def visual_temperature_step(value):
+    VISUAL_TEMPERATURE_STEP_SCHEMA = cv.Schema(
+        {
+            cv.Required(CONF_TARGET_TEMPERATURE): visual_temperature,
+            cv.Required(CONF_CURRENT_TEMPERATURE): visual_temperature,
+        }
+    )
```
Comment (jesserockz) [inline esphome/components/climate/__init__.py]: The schema should be outside the function as we dont want to recreate it multiple times
```suggestion
VISUAL_TEMPERATURE_STEP_SCHEMA = cv.Schema(
    {
        cv.Required(CONF_TARGET_TEMPERATURE): visual_temperature,
        cv.Required(CONF_CURRENT_TEMPERATURE): visual_temperature,
    }
)

def visual_temperature_step(value):
```

**Review by jesserockz — DISMISSED** [review]

File: esphome/components/climate/__init__.py (hunk 1 above)
Comment (hostcc) [reply esphome/components/climate/__init__.py]: Thanks, Jesse - makes sense!

**Comment by hostcc** [comment]
> This is what I mean:
> 
> ```python
> VISUAL_TEMPERATURE_STEP_SCHEMA = cv.Schema(
>         {
>             cv.Required(CONF_TARGET_TEMPERATURE): visual_temperature,
>             cv.Required(CONF_CURRENT_TEMPERATURE): visual_temperature,
>         }
>     )
> 
> def visual_temperature_steps(value):
>     if isinstance(value, dict):
>         return VISUAL_TEMPERATURE_STEP_SCHEMA(value)
> 
>     return VISUAL_TEMPERATURE_STEP_SCHEMA(
>         {
>             CONF_TARGET_TEMPERATURE: value,
>             CONF_CURRENT_TEMPERATURE: value,
>         }
>     )
> 
>                     cv.Optional(CONF_TEMPERATURE_STEP): visual_temperature_steps,
> ```
> 
> It's a quick copy and paste and will need indent fixing, but that's the idea.

Thanks for the idea, Samuel! I took it into the implementation.

**Review by ssieb — APPROVED** [review]

**Pull request merged**